src/app/ryderserial/ryder-serial/index.ts

```diff
@@ -140,359 +131,345 @@ export default class RyderSerial extends Events.EventEmitter {
-   * - close connection,
-   * - clear reconnect interval
-   * - destroy `this.serial`
+  /*
+   * Schedules repeated attempts to reconnect to the bridge, separated by a delay of
+   * `this.options.reconnect_interval_ms`.
    */
-  public close(): void {
-    if (this.closing) {
-      return;
-    }
-    this.closing = true;
-    this.clear(); // clears watchdog timeout, rejects all pending processes, and releases all locks.
-    this.connection?.close(); // close connection if it exists
-    clearInterval(this[reconnect_symbol]); // clear reconnect interval
-    delete this.connection; // destroy connection
+  private set_reconnect_interval() {
+    let delay = this.options.reconnect_time || 1000;
+    this.log(LogLevel.DEBUG, `Retrying connection in ${delay} ms`);
+    // Clear the previous timer
+    clearInterval(this.reconnect_timer);
+    // Create a new one
+    this.reconnect_timer = setInterval(
+      this.connect_to_bridge.bind(this),
+      delay,
+    );
   }
 
-  /**
-   * @returns `true` if RyderSerial is currently locked; `false` otherwise
+  /*
+   * Handles the WebSocket connection opening successfully. Note that this does not mean that the
+   * `RyderSerial` is ready for use, because the bridge may later indicate that another client
+   * already has control of the device.
    */
-  public locked(): boolean {
-    return !!this[lock_symbol].length;
+  private on_open() {
+    this.log(LogLevel.DEBUG, 'WS connection opened');
+    this.set_state(State.WAITING_FOR_BRIDGE_STATE_RESPONSE);
+    // Remove any reconnection timer because the connection is now open
+    clearInterval(this.reconnect_timer);
   }
 
-  /**
-   * Requests a lock to be placed so that commands can be sent in sequence.
-   *
-   * @returns a `Promise` that resolves when the lock is released.
+  /* Handles errors encountered while interacting with the Ryder bridge or device. */
+  private on_error(error: Error) {
+    // Note that `on_close` will be called on error as well
+    this.log(LogLevel.ERROR, `Error in Ryder bridge or device connection: ${error}`);
+    // This causes an uncaught error, but 'failed' seems to be caught correctly
+    // this.emit('error', error);
+    this.emit('failed', error);
+    // Clear the command queue
+    this.queue.cancel_all();
+    // Close the bridge connection if it isn't already closed
+    this.close_connection();
+    // Retry the connection after a delay
+    this.set_reconnect_interval();
+  }
+
+  /* Handles the connection closing for any reason. */
+  private on_close() {
+    this.log(LogLevel.DEBUG, 'WS connection closed');
+    // Only update the state if the `RyderSerial` is not permanently closed
+    if (this.state !== State.CLOSED_PERMANENTLY) {
+      this.set_state(State.CLOSED);
+    }
+    this.emit('close');
+    // Clear the command queue
+    this.queue.cancel_all();
+  }
+
+  /* Handles responses received from the Ryder bridge and device. */
+  private on_response(response: any) {
```

Comment:
type is `string | UInt8Array` ?

Reply:
And ideally just `Uint8Array` 😁

Reply:
Strangely, it seems that it is actually `string | object`, not `string | UInt8Array`. I had to convert to a `UInt8Array` later on in `on_raw_device_response`.